python/dgl/distributed/graph_partition_book.py

```diff
@@ -335,6 +335,261 @@ def get_edge_size(self):
+    def __init__(self, part_id, num_parts, node_map, edge_map):
+        assert part_id >= 0, 'part_id cannot be a negative number.'
+        assert num_parts > 0, 'num_parts must be greater than zero.'
+        self._partid = part_id
+        self._num_partitions = num_parts
+        node_map = utils.toindex(node_map)
+        edge_map = utils.toindex(edge_map)
+        self._node_map = node_map.tonumpy()
+        self._edge_map = edge_map.tonumpy()
+        # Get meta data of GraphPartitionBook
+        self._partition_meta_data = []
+        for partid in range(self._num_partitions):
+            nrange_start = node_map[partid - 1] if partid > 0 else 0
+            nrange_end = node_map[partid]
+            erange_start = edge_map[partid - 1] if partid > 0 else 0
+            erange_end = edge_map[partid]
+            part_info = {}
+            part_info['machine_id'] = partid
+            part_info['num_nodes'] = int(nrange_end - nrange_start)
+            part_info['num_edges'] = int(erange_end - erange_start)
+            self._partition_meta_data.append(part_info)
+
+    def shared_memory(self, graph_name):
+        """Move data to shared memory.
+
+        Parameters
+        ----------
+        graph_name : str
+            The graph name
+        """
+        self._meta = _move_metadata_to_shared_mam(
+            graph_name, self._num_nodes(), self._num_edges(), self._partid,
+            self._num_partitions, F.tensor(self._node_map), F.tensor(self._edge_map), True)
+
+    def num_partitions(self):
+        """Return the number of partitions.
+
+        Returns
+        -------
+        int
+            number of partitions
+        """
+        return self._num_partitions
+
+
+    def _num_nodes(self):
+        """ The total number of nodes
+        """
+        return int(self._node_map[-1])
+
+    def _num_edges(self):
+        """ The total number of edges
+        """
+        return int(self._edge_map[-1])
+
+    def metadata(self):
+        """Return the partition meta data.
+
+        The meta data includes:
+
+        * The machine ID.
+        * The machine IP address.
+        * Number of nodes and edges of each partition.
+
+        Examples
+        --------
+        >>> print(g.get_partition_book().metadata())
+        >>> [{'machine_id' : 0, 'num_nodes' : 3000, 'num_edges' : 5000},
+        ...  {'machine_id' : 1, 'num_nodes' : 2000, 'num_edges' : 4888},
+        ...  ...]
+
+        Returns
+        -------
+        list[dict[str, any]]
+            Meta data of each partition.
+        """
+        return self._partition_meta_data
+
+
+    def nid2partid(self, nids):
+        """From global node IDs to partition IDs
+
+        Parameters
+        ----------
+        nids : tensor
+            global node IDs
+
+        Returns
+        -------
+        tensor
+            partition IDs
+        """
+        nids = utils.toindex(nids)
+        ret = np.searchsorted(self._node_map, nids.tonumpy(), side='right')
```

Comment:
Will searchsorted() bring much more overhead than a slice? This API would be invoked very frequent.